File: C-to-Wasm/revenue_distribution/Optimized/DistributionContract.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include "hashmap.h"
#include "DistributionContract.h"

#define INDEX_INIT 1000000
#define INCR_PER_REV_INIT 10000

// Utility function prototype
void distributeRevenue(DistributionContract* contract, double amount);

// Hash function
uint64_t userDataHash(const void* item, uint64_t seed0, uint64_t seed1) {
    const UserState* userState = item;
    return hashmap_sip(userState->address, strlen(userState->address), seed0, seed1);
}

// Compare function
int userDataCompare(const void* a, const void* b, void* udata) {
    const UserState* userState1 = a;
    const UserState* userState2 = b;
    return strcmp(userState1->address, userState2->address);
}

/**
 * @brief Constructs a distribution contract
 * 
 * @return DistributionContract*
 */
DistributionContract* constructContract() {
    DistributionContract* contract = calloc(1, sizeof(DistributionContract));
    if(!contract) return NULL;
    contract->userStateMap = hashmap_new(sizeof(UserState), 0, 0, 0, userDataHash, userDataCompare, NULL, NULL);
    if(!contract->userStateMap) {
        free(contract);
        return NULL;
    }
    contract->totalStake = 0;
    contract->incrementPerRevenue = INCR_PER_REV_INIT;
    contract->index = INDEX_INIT;
    return contract;
}

/**
 * @brief Destroys a distribution contract
 * 
 * @param contract The contract to be destroyed
 */
void destroyContract(DistributionContract* contract) {
   // Sanitization
   if(!contract) return;
   hashmap_free(contract->userStateMap);
   free(contract);
}

/**
 * @brief Function to add share to the destination address
 * WARNING: Any user may add as much share as they want
 * This has been done to isolate only the revenue distribution
 * and not the transfer (whose time can vary depending on the implementation)
 * 
 * @param contract The contract 
 * @param dest The destination address of the change
 * @param change The amount to add/remove
 * @return int Success code: 0 if succeeded else transaction revert
 */
int changeShare(DistributionContract* contract, char* dest, double change) {
    // Sanity checks
    if(change == 0 || !contract || !dest) return EXIT_FAILURE;
    double newTotalStake = contract->totalStake + change;
    double oldTotalStake = contract->totalStake;
    if(newTotalStake < 0 || !isfinite(newTotalStake)) return EXIT_FAILURE;
    // Get user data
    UserState* userState = hashmap_get(contract->userStateMap, &(UserState){ .address = dest });
    UserState tmp = { 0 };
    if(!userState) {
        // If no mapping exists
        tmp = (UserState){ dest, 0, contract->totalStake, contract->incrementPerRevenue, 0, contract->index };    
    } else {
        // If there was a mapping
        tmp = *userState;
    }
    if(tmp.ownStake + change < 0) return EXIT_FAILURE;
    // Phase 1
    double freshOwn = tmp.ownStake == 0 ? 0 : (contract->index - tmp.lastIndex) * tmp.ownStake / 
        (tmp.lastIncrementPerRevenue * tmp.lastTotalStake);
    tmp.ownAccumulatedTotal += freshOwn;
    // Phase 2
    if(newTotalStake != 0) { // TODO: Floating points can be dangerous!
        contract->incrementPerRevenue = INCR_PER_REV_INIT;
    } else if(oldTotalStake != 0) {
        contract->incrementPerRevenue *= oldTotalStake / newTotalStake;
    }
    contract->totalStake = newTotalStake;
    // Phase 3
    tmp.ownStake += change;
    tmp.lastIndex = contract->index;
    tmp.lastIncrementPerRevenue = contract->incrementPerRevenue;
    tmp.lastTotalStake = contract->totalStake;
    /// Update
    hashmap_set(contract->userStateMap, &tmp);
    return EXIT_SUCCESS;
}

/**
 * @brief Injects revenue into the contract
 * 
 * @param contract The destination address
 * @param amount The amount to add 
 * @return int Success code: 0 if succeeded else transaction revert
 */
int addRevenue(DistributionContract* contract, double amount) {
    // Sanity checks
    if(!contract || amount <= 0) return EXIT_FAILURE;
    distributeRevenue(contract, amount);
    return EXIT_SUCCESS;
}

/**
 * @brief Utility distribution function
 *
 * @param contract The destination address
 * @param amount The amount to distribute
 */
void distributeRevenue(DistributionContract* contract, double amount) {
    // Efficient distribution
    contract->index += contract->incrementPerRevenue * amount;
}

File: Java/src/main/java/src/revenue_distribution/optimized/DistributionContract.java
package src.revenue_distribution.optimized;

import src.TxnRevertException;

import java.util.HashMap;
import java.util.Map;

import static src.TxnRevertException.assertOrRevert;

/**
 * A simple naive revenue distribution contract
 */
public final class DistributionContract {

    // Global arbitrary initialization constants
    private final static double INDEX_INIT = 1000000;
    private final static double INCR_PER_REV_INIT = 10000;

    /**
     * Basic struct that models all the data associated to one address
     */
    private static final class UserData {
        private double ownStake;
        private double lastTotalStake;
        private double lastIncrementPerRevenue;
        private double ownAccumulatedTotal;
        private double lastIndex;
        public UserData(GlobalData globalData) {
            this(globalData, 0);
        }
        public UserData(GlobalData globalData, double share) {
            this.ownStake = share;
            this.lastTotalStake = globalData.totalStake;
            this.lastIncrementPerRevenue = globalData.incrementPerRevenue;
            this.ownAccumulatedTotal = 0;
            this.lastIndex = globalData.index;
        }
    }

    /**
     * Basic struct that models all global data of the contract
     */
    private static final class GlobalData {
        private double totalStake;
        private double incrementPerRevenue;
        private double index;
        public GlobalData() {
            this.totalStake = 0;
            this.incrementPerRevenue = INCR_PER_REV_INIT;
            this.index = INDEX_INIT;
        }
    }

    // Contract's internal storage
    private final Map<String, UserData> userDataMap;
    private final GlobalData globalData;


    /**
     * Contract's constructor, for personal testing setup
     * @param map The address -> share mapping
     */
    public DistributionContract(Map<String, Integer> map) {
        var userDataMap = new HashMap<String, UserData>();
        var globalData = new GlobalData();
        // First loop to create global data
        for(var value : map.values()) {
            assertOrRevert(value > 0);
            globalData.totalStake += value;
        }
        // Second loop to create user data, once global data is up-to-date
        for(var entry : map.entrySet()) {
            userDataMap.put(entry.getKey(), new UserData(globalData, entry.getValue()));
        }
        this.userDataMap = userDataMap;
        this.globalData = globalData;
    }

    /**
     * Public function to add share to the sender
     * WARNING: Any user may add as much share as they want
     * This has been done to isolate only the revenue distribution
     * and not the transfer (whose time can vary depending on the implementation)
     * @param dest The destination address of the change
     * @param change The amount to add/remove
     * @exception TxnRevertException if the change incurs an over/underflow
     */
    public void changeShare(String dest, double change) {
        // Sanitization
        assertOrRevert(change != 0);
        var newTotalStake = globalData.totalStake + change;
        var oldTotalStake = globalData.totalStake;
        assertOrRevert(Double.isFinite(newTotalStake) && newTotalStake >= 0);
        UserData userData = userDataMap.getOrDefault(dest, new UserData(globalData));
        assertOrRevert(userData.ownStake + change >= 0);
        // Phase 1
        var freshOwn = userData.ownStake == 0 ? 0 : (globalData.index - userData.lastIndex) * userData.ownStake /
                (userData.lastIncrementPerRevenue * userData.lastTotalStake);
        userData.ownAccumulatedTotal += freshOwn;
        // Phase 2
        if(newTotalStake != 0) { // TODO: Floating points CAN be dangerous!
            globalData.incrementPerRevenue = INCR_PER_REV_INIT;
        } else if(oldTotalStake != 0) {
            globalData.incrementPerRevenue *= oldTotalStake / newTotalStake;
        }
        globalData.totalStake = newTotalStake;
        // Phase 3
        userData.ownStake += change;
        userData.lastIndex = globalData.index;
        userData.lastIncrementPerRevenue = globalData.incrementPerRevenue;
        userData.lastTotalStake = globalData.totalStake;
        userDataMap.put(dest, userData);
    }

    /**
     * Injects revenue into the contract
     * @param amount The amount to add
     * @exception TxnRevertException if amount is not strictly positive
     */
    public void addRevenue(double amount) {
        // Check whether the change is strictly positive
        assertOrRevert(amount > 0);
        distributeRevenue(amount);
    }

    /**
     *  Utility distribution function
     * @param amount The amount to distribute
     */
    private void distributeRevenue(double amount) {
        // Efficient distribution
        globalData.index += globalData.incrementPerRevenue * amount;
    }
}